sklearn/utils/_encode.py

```diff
@@ -273,3 +315,53 @@ def is_valid(value):
     if return_mask:
         return diff, valid_mask
     return diff
+
+
+class _NaNCounter(Counter):
+    """Counter with support for nan values."""
+
+    def __init__(self, items):
+        super().__init__(self._generate_items(items))
+
+    def _generate_items(self, items):
+        """Generate items without nans. Stores the nan counts seperately."""
+        for item in items:
+            if not is_scalar_nan(item):
+                yield item
+                continue
+            if not hasattr(self, "nan_count"):
+                self.nan_count = 0
+            self.nan_count += 1
+
+    def __missing__(self, key):
+        if hasattr(self, "nan_count") and is_scalar_nan(key):
+            return self.nan_count
+        raise KeyError(key)
+
+
+def _get_counts(values, uniques):
+    """Get the count of each of the `uniques` in `values`. The counts will use
+    the order passed in by `uniques`.
```

Comment:
If a category is in `values` but not in `uniques`, it will be ignored:

```python
from sklearn.utils._encode import _get_counts
import numpy as np

_get_counts(np.asarray([1, 2, 3, 4, 4, 4]), np.asarray([1, 4]))
# array([1, 3])
```